File: palindromic-substrings/njngwn.java
// Time Complexity: O(n*n), n: s.length()
// Space Complexity: O(1)
class Solution {
    public int countSubstrings(String s) {
        int cnt = 0;

        for (int i = 0; i < s.length(); i++, cnt++) {   // i: center of palindrom
            // odd number palindrom: e.g. aba
            int left = i-1, right = i+1;

            while (left >= 0 && right < s.length()) {
                if (s.charAt(left) != s.charAt(right)) {
                    break;
                }
                cnt++;
                left--;
                right++;
            }

            // even number palindrom: e.g. abba
            left = i;
            right = i+1;

            while (left >= 0 && right < s.length()) {
                if (s.charAt(left) != s.charAt(right)) {
                    break;
                }
                cnt++;
                left--;
                right++;
            }
        }

        return cnt;
    }
}

File: palindromic-substrings/njngwn.py
class Solution:
    # Time Complexity: O(n^2), n: len(s)
    # Space Complexity: O(1)
    def countSubstrings(self, s: str) -> int:
        cnt = 0

        for i in range(len(s)):
            # odd number
            start, end = i, i
            while start >= 0 and end < len(s) and s[start] == s[end]:
                start -= 1
                end += 1
                cnt += 1

            # even number
            start, end = i, i + 1
            while start >= 0 and end < len(s) and s[start] == s[end]:
                start -= 1
                end += 1
                cnt += 1

        return cnt
